plugins/epochstorage/plugin.go

```diff
@@ -0,0 +1,474 @@
+	_ = epochContentStorage.messageIDs.IterateKeys(kvstore.EmptyPrefix, func(key kvstore.Key) bool {
+		var msgID tangle.MessageID
+		if _, err := msgID.Decode(key); err != nil {
+			panic("MessageID could not be parsed!")
+		}
+		messageIDs = append(messageIDs, msgID)
+		return true
+	})
+	return messageIDs, nil
+}
+
+func insertMessageToEpoch(ei epoch.Index, msgID tangle.MessageID) error {
+	epochContentStorage, err := getEpochContentStorage(ei)
+	if err != nil {
+		return err
+	}
+
+	if err := epochContentStorage.messageIDs.Set(msgID.Bytes(), msgID.Bytes()); err != nil {
+		return errors.New("Fail to insert Message to epoch store")
+	}
+	return nil
+}
+
+func removeMessageFromEpoch(ei epoch.Index, msgID tangle.MessageID) error {
+	epochContentStorage, err := getEpochContentStorage(ei)
+	if err != nil {
+		return err
+	}
+
+	if err := epochContentStorage.messageIDs.Delete(msgID.Bytes()); err != nil {
+		return errors.New("Fail to remove Message from epoch store")
+	}
+	return nil
+}
+
+func insertTransactionToEpoch(ei epoch.Index, txID utxo.TransactionID) error {
+	epochContentStorage, err := getEpochContentStorage(ei)
+	if err != nil {
+		return err
+	}
+
+	if err := epochContentStorage.transactionIDs.Set(txID.Bytes(), txID.Bytes()); err != nil {
+		return errors.New("Fail to insert Transaction to epoch store")
+	}
+	return nil
+}
+
+func removeTransactionFromEpoch(ei epoch.Index, txID utxo.TransactionID) error {
+	epochContentStorage, err := getEpochContentStorage(ei)
+	if err != nil {
+		return err
+	}
+
+	if err := epochContentStorage.transactionIDs.Delete(txID.Bytes()); err != nil {
+		return errors.New("Fail to remove Transaction from epoch store")
+	}
+	return nil
+}
+
+func insertOutputsToEpoch(ei epoch.Index, spent, created []*ledger.OutputWithMetadata) error {
+	epochContentStorage, err := getEpochContentStorage(ei)
+	if err != nil {
+		return err
+	}
+
+	for _, s := range spent {
+		if err := epochContentStorage.spentOutputs.Set(s.ID().Bytes(), s.ID().Bytes()); err != nil {
+			return errors.New("Fail to insert spent output to epoch store")
+		}
+	}
+
+	for _, c := range created {
+		if err := epochContentStorage.createdOutputs.Set(c.ID().Bytes(), c.ID().Bytes()); err != nil {
+			return errors.New("Fail to insert created output to epoch store")
+		}
+	}
+
+	return nil
+}
+
+func removeOutputsFromEpoch(ei epoch.Index, spent, created []*ledger.OutputWithMetadata) error {
+	epochContentStorage, err := getEpochContentStorage(ei)
+	if err != nil {
+		return err
+	}
+
+	for _, s := range spent {
+		if err := epochContentStorage.spentOutputs.Delete(s.ID().Bytes()); err != nil {
+			return errors.New("Fail to remove spent output from epoch store")
+		}
+	}
+
+	for _, c := range created {
+		if err := epochContentStorage.createdOutputs.Delete(c.ID().Bytes()); err != nil {
+			return errors.New("Fail to remove created output from epoch store")
+		}
+	}
+
+	return nil
+}
+
+func saveEpochVotersWeight(message *tangle.Message) {
```

Comment:
If a voter already voted for another ECR of the same epoch we should delete the old entry! To determine which vote is the newest you will have to save the message IssuingTimestamp